Nvpmodel snap integration

EstebanVg15 opened the pull request:
This snap has been tested on Jetson Orin Nano/NX and AGX Orin devices running Ubuntu Core.

EstebanVg15 committed:
Create a preliminary version of the nvpmodel snap

EstebanVg15 committed:
Add the correct devices in sys to read and set the NVPMODEL

EstebanVg15 committed:
Stage the coreutils package to have the whoami command on the build needed by jetson-clocks

EstebanVg15 committed:
Add sysfs for jetson-clocks to read the system data and only provide the jetson-clocks --show command

EstebanVg15 committed:
Stage systemd to use the systemctl command

EstebanVg15 committed:
Add the necessary interfaces for nvpmodel to reboot a Jetson device after switching a power management mode

EstebanVg15 committed:
Add a preliminary README file for the nvpmodel snap

EstebanVg15 committed:
Provide nvfancontrol as a service for the nvpmodel snap

jetson_clocks runs the following commands
before reading the clocks and the fan status.
    systemctl is-active nvfancontrol
​    systemctl is-active nv-spi-server

Unfortunately running systemctl from a snap is not possible.
Alternatively snapctl can be used.
    https://forum.snapcraft.io/t/enable-disable-systemd-services/10626/2

So the nvidia-l4t-nvfancontrol stage package has been included,
and usr/sbin/nvfancontrol has been set as a daemon to be monitored by snapctl.

Finally a bash script is executed instead of "systemctl is-active" to run "snapctl services"
to retrieve the status of nvfancontrol.

Important notes:
1. Currently the nvfancontrol service cannot be fully started since the snap does not have access to /var/lock/nvf.lock,
   However this functionality does not impact our use case.
2. Checking the nv-spi-server service status is currently being bypassed, similarly it does not impact our use case.

EstebanVg15 committed:
Update the nvpmodel README file removing some runtime errors that were present

EstebanVg15 committed:
Nvpmodel: Revert the workaround of using nvfancontrol and a custom systemctl implementation

EstebanVg15 committed:
Nvpmodel: Define the corresponding services for the nvpmodel snap to be properly configured

 * Bind /etc/nvpmodel as a layout for the snap to access all the available configuration files.

 * Set etc/systemd/nvpower.sh as a service to read the CPU specs and determine which
   configuration file should be set as /etc/nvpmodel.conf

 * Create service for nvpmodel which initializes the environment with a default configuration.

 * Declare the remaining system-files for the services to access the required information

EstebanVg15 committed:
Nvpmodel: Add remaining system-files for the nvpower-service to be started successfully

EstebanVg15 committed:
Nvpmodel: Update the README file of the nvpmodel snap

EstebanVg15 committed:
Nvpmodel: Add system-files for the nvpmodel snap to work on Jetson Orin NX devices

EstebanVg15 committed:
Nvpmodel: Define a start order for the nvpower and the nvpmodel services

EstebanVg15 committed:
Nvpmodel: Update the README including extra configuration steps when installing the snap

Restarting the snap is needed because inside the snap there are two services (nvpower and nvpmodel)
which need the set-power-management-mode interface to be connected to be properly executed.

EstebanVg15 committed:
Nvpmodel: Add system-files for the nvpmodel snap to work on Jetson AGX Orin devices

EstebanVg15 committed:
Nvpmodel: Add a brief description of the nvpmodel snap on the base README file

DocSepp commented on nvpmodel/README.md:
```diff
@@ -0,0 +1,68 @@
+# nvpmodel
+
+This snap packages the `nvpmodel` and the `jetson_clocks` utilities which can be used to set a determined power mode (e.g., `7W`, `15W`, `25W`, `MAXN_SUPER`, etc...) and read the mode that has configured, displaying the system clocks (CPU, GPU, and EMC) and the fan status of a Jetson device.
```
"read the mode that was configured"

EstebanVg15 replied:
Thanks for pointing that out! I didn't see that mistake.

DocSepp commented on nvpmodel/README.md:
```diff
@@ -0,0 +1,68 @@
+
+This snap doesn't require any extra build steps for the software it provides since it uses staging packages, such as some specific NVIDIA L4T packages (i.e., `nvidia-l4t-nvpmodel`, `nvidia-l4t-core` and `nvidia-l4t-tools`) to provide runtime libraries and binaries.
+
+Some `layouts` were defined to bind-mount directories from inside the snap to a different location in the execution environment. This is necessary because the tools expect certain libraries and configuration files in hard-coded paths by default, however, all files included in the snap are stored under `/snap/nvpmodel/current`.
```
I think you can leave out the last part because the files of a snap are always stored in `/snap/snap-name/current`.

EstebanVg15 replied:
Understood, I agree with you.

DocSepp commented on nvpmodel/README.md:
```diff
@@ -0,0 +1,68 @@
+
+# Build, install and run the snap.
+
+This snap has no external dependencies for being built and installed, there is just needed to run `snapcraft` from the `nvpmodel/` directory, install the resulting snap using the `--dangerous` flag, connect all necessary interfaces, and finally restart the snap, this last step is needed because inside the snap there are two services (i.e., `nvpower` and `nvpmodel`) which need the set-power-manage ...
```
"you just need to run `snapcraft`"

I think mentioning the `nvpmodel` directory here can be confusing because the README is already in that directory. It could make you think that there's a `nvpmodel/nvpmodel` directory.

EstebanVg15 replied:
Understood, I agree with you.

DocSepp commented on nvpmodel/README.md:
```diff
@@ -0,0 +1,68 @@
+
+# Update and read the power management mode.
+
+The command shown below can be used to update the power management mode of a `Jetson Orin Nano 8GB` device, `-m 0` corresponds on this case to `15W`, when switching between power modes the snap might request the system to be rebooted.
```
It won't always reboot the system after setting the power mode?

EstebanVg15 replied:
No, sometimes the tool asks the user to reboot the device, and other times it just update the Power Mode, I will do some research to identify which are the circumstances for a system reboot to be requested.

DocSepp commented on nvpmodel/README.md:
````diff
@@ -0,0 +1,68 @@
+$ sudo snap run nvpmodel.nvpmodel -m 0
+```
+
+The following command can be used to retrieve the power management mode that is currently set on a  `Jetson Orin Nano 8GB` device.
````
This should work on any Tegra device, shouldn't it?

EstebanVg15 replied:
Yes, --query retrieves the operating mode for every device, in previous versions of this snap, I used to set as part of the command the configuration file needed for the Jetson Orin Nano 8GB device, and there was where this part of the README made more sense, but now that the corresponding service creates the symlink on /etc/nvpmodel.conf there is no need to provide the corresponding configuration file since it is auto selected, I will update this part of the README.

DocSepp commented on nvpmodel/README.md:
```diff
@@ -0,0 +1,68 @@
+
+As can be seen on the example above, the settings corresponds to the `15W` power mode that was set by `nvpmodel`.
+
+[Supported Modes and Power Efficiency](https://docs.nvidia.com/jetson/archives/r36.4.3/DeveloperGuide/SD/PlatformPowerAndPerformance/JetsonOrinNanoSeriesJetsonOrinNxSeriesAndJetsonAgxOrinSeries.html#sd-platformpowerandperformance-supportedmodesandpowerefficiency)
```
You already linked to this in line 7

DocSepp commented on nvpmodel/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,170 @@
+
+layout:
+  /etc/nvpmodel:
+    bind: $SNAP/etc/nvpmodel
```
Should the nvpmodel configuration be modifiable manually? In that case I guess you should bind `$SNAP_DATA/etc/nvpmodel` instead and install it using an install hook.

In that case, you shouldn't update it using a post-refresh hook though because it might overwrite the manually changed configuration file. We should probably talk about this use case.

EstebanVg15 replied:
Hi @DocSepp ,

As per my understanding the configuration files should not be modified manually.

The `/etc/nvpmodel/`  is a directory that contains the configuration files for each hardware variant as shown below.

```
root@ubuntu:/home/esteban-vazquez# ls /etc/nvpmodel/
nvpmodel_igx_orin.conf	       nvpmodel_p3701_0008.conf        nvpmodel_p3767_0003.conf        nvpmodel_t194_8gb.conf	      nvpmodel_t194_p3668_emul.conf
nvpmodel_igx_orin_safety.conf  nvpmodel_p3767_0000.conf        nvpmodel_p3767_0003_super.conf  nvpmodel_t194_agxi.conf
nvpmodel_p3701_0000.conf       nvpmodel_p3767_0000_super.conf  nvpmodel_p3767_0004.conf        nvpmodel_t194_e3900_dGPU.conf
nvpmodel_p3701_0002.conf       nvpmodel_p3767_0001.conf        nvpmodel_p3767_0004_super.conf  nvpmodel_t194_e3900_iGPU.conf
nvpmodel_p3701_0004.conf       nvpmodel_p3767_0001_super.conf  nvpmodel_t194.conf	       nvpmodel_t194_p3668.conf
```

Each file contains the supported power modes and the parameters that must be set for each mode, these files are provided within the tool installation directory of the snap and them are Read Only that's why I am taking those from `$SNAP/etc/nvpmodel/`, finally the `etc/systemd/nvpower.sh` service read some CPU properties and settings to identify the current HW variant and automatically creates a symlink from one of these `/etc/nvpmodel/*.conf` files to -> `/etc/nvpmode.conf` .

DocSepp replied:
yeah you're right, we shouldn't be able to modify those by hand

DocSepp commented on nvpmodel/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,170 @@
+apps:
+  nvpower-service:
+    command-chain: [bin/wrapper_local]
+    command: etc/systemd/nvpower.sh
```
This service will create the link from `/etc/nvpmodel.conf` to `/etc/nvpmodel/nvpmodel_<platform>.conf`. 

Theoretically, we could bind-mount `$SNAP_DATA/nvpmodel.conf` to `/etc/nvpmodel.conf` to avoid having to access this through system-files.

EstebanVg15 replied:
Hi @DocSepp ,

I initially attempted to create a bind-file/layout on `$SNAP_DATA/etc/nvpmodel.conf` however this creates an empty file, and moreover during the `etc/systemd/nvpower.sh` execution the services checks if `etc/nvpmodel.conf` exists, and as it exists as an empty file (due to the layout creation), the service skips the part of reading the HW variant and creating the corresponding symlink.

That's why I ended using `system-files` instead.

DocSepp replied:
Right, that makes sense with the bind-mount. Have you tried it with a symlink? I assume that it would be different in that case. However, the symlink won't work, if that file is already present on the filesystem as on Classic for example.

I think it's fine to leave the system-files access to that file.

EstebanVg15 replied:
Thanks for the feedback on this!

Yes, I tried to use a symlink layout property, but on this case the nvpower service failed to create the proper symlink on runtime.

I also tested removing the layouts properties for /etc/nvpmodel.conf and creating a symlink as an install hook, this partially worked but the symbolic link was static and could not be updated by the snap on runtime when using a different HW variant.

That's why I think that using the system-files interface is the best way for the service to detect the HW variant and create the corresponding symbolic link on runtime.

DocSepp replied:
That sounds reasonable, thanks for confirming.

DocSepp commented on nvpmodel/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,170 @@
+  hardware-observe:
+  set-power-management-mode:
+    interface: system-files
+    write:
```
Do you really need write access to all these files because the hardware-observe interface should already give read access to all the devices in `/sys/devices`.

EstebanVg15 replied:
Hi @DocSepp ,

I've been checking this today, and most of the files need write access since the `etc/systemd/nvpower.sh` and the `usr/sbin/nvpmodel` tools store information on them.

The files below seem to be the only ones that just need read access, and those are used by `usr/sbin/nvpmodel` and `usr/bin/jetson_clocks` however, the `hardware-observe` interface does not seem to be enough for them, since even though with such interface defined in the `plugs` section of each app and being connected on runtime, I still get the errors below.

```
cat: /sys/kernel/nvpmodel_clk_cap/emc: Permission denied
cat: /sys/kernel/debug/bpmp/debug/clk/emc/min_rate: Permission denied
cat: /sys/kernel/nvpmodel_clk_cap/emc: Permission denied
cat: /sys/kernel/debug/bpmp/debug/clk/emc/rate: Permission denied
cat: /sys/kernel/debug/bpmp/debug/clk/emc/mrq_rate_locked: Permission denied
```

This seems to be fixed only when using `system-files`, so, I can define a `system-file read-only` entry for those, just for the snap to be completely specific.

EstebanVg15 replied:
Just a minor comment, after testing the snap on a Jetson Orin Nano/NX, and a Jetson AGX I saw that sometimes it is needed to write to the /sys/kernel/nvpmodel_clk_cap/emc file so I set write permissions for it, the other the EMC related files are marked as read only.

DocSepp replied:
That's alright, no worries. I was mostly just curious :)

I also saw that the hardware observe interface doesn't grant access to `/sys/kernel/debug/bpmp`. In case you didn't know, you can just see the apparmor profiles generated for an interface by checking the source code: https://github.com/canonical/snapd/tree/master/interfaces/builtin

DocSepp commented on nvpmodel/snap/snapcraft.yaml:
```diff
@@ -0,0 +1,170 @@
+      - /sys/devices/system/cpu/cpufreq/policy0/scaling_min_freq
+      - /sys/devices/system/cpu/cpufreq/policy4/scaling_min_freq
+      - /sys/devices/system/cpu/cpufreq/policy8/scaling_min_freq
+      - /sys/devices/system/cpu/cpufreq/schedutil/rate_limit_us
```
You could simplify this by just specifying `/sys/devices/system/cpu`. The resulting apparmor profile will give rw access to all subdirectories. As you can see in `/var/lib/snapd/apparmor/profiles/snap.nvpmodel.nvpmodel-service`:
```
"/sys/devices/system/cpu{,/,/**}" rwkl,
```
I think that would keep this easier to manage. Also, you are alredy giving write access to so many files, so I don't think it's that big of a deal to include a few more and making our life easier...

EstebanVg15 replied:
Sure thanks for the suggestion, I implemented the corresponding modification.

DocSepp reviewed (commented):
Hey Esteban, thanks a lot for these changes!

I left a couple of suggestions but nothing major :)

EstebanVg15 committed:
Nvpmodel: Update the README file of the nvpmodel snap for it to be more clear and remove unneeded references

EstebanVg15 committed:
Nvpmodel: Simplify the system-files interface definition on the nvpmodel snap

 * Grant access to /sys/devices/system/cpu instead of defining each one of the corresponding sub entries
 * Grant read-o access to the EMC sys file systems since the tools does not store information in there

EstebanVg15 committed:
Nvpmodel: Grant write access to some EMC sysfs

Even though when testing the snap on a Jetson Orin Nano this sysfs was only accessed
With read permissions on some cases for the Jetson Orin NX there is needed to write to this entry

DocSepp commented on nvpmodel/snap/snapcraft.yaml:
```diff
@@ -64,10 +47,6 @@ plugs:
       - /sys/devices/platform/17000000.gpu/devfreq_dev/max_freq
       - /sys/class/devfreq/17000000.gpu/device/of_node/compatible
       - /sys/devices/platform/bus@0/17000000.gpu/railgate_enable
```
I also think you should group all these together, as well as some of the nodes that come after.

Also note that `/sys/class/devfreq/17000000.gpu` and `/sys/devices/platform/17000000.gpu` are just symbolic links to `/sys/devices/platform/bus@0/17000000.gpu`. 

I think you should be able to get away with just granting access to the real path. That would also make it more consistent.

EstebanVg15 replied:
Hi @DocSepp ,

Thanks for the suggestion, as per today's call I modified the `system-files` list to use some directories instead of individual files as well as real directories instead of symbolic links for the list to be simplified. additionally, I tested the changes on a Jetson Orin Nano/NX and on an AGX Orin device.

DocSepp commented on nvpmodel/README.md:
````diff
@@ -33,7 +33,7 @@ The command shown below can be used to update the power management mode of a `Je
 ```
 
-The following command can be used to retrieve the power management mode that is currently set on a  `Jetson Orin Nano 8GB` device.
+The following command can be used to retrieve the power management mode that is currently a device.
````
It should still be "set on a device" :P

EstebanVg15 replied:
Thanks for pointing that out! I've fixed the README accordingly.

DocSepp reviewed (commented):
Thanks again for the improvements!

It's probably not that significant to separate the three nodes that only need read-access but I think it looks better :D

I just think you should group some more of these nodes together.

EstebanVg15 committed:
Nvpmodel: Fix typo on the nvpmodel snap README

EstebanVg15 committed:
Nvpmodel: Simplify the system-files list granting access to the complete folder instead of individual files

DocSepp reviewed (approved):
Thanks for the changes!

Looks good to me now.
